File: Sources/AppShare/AppShareViewController.m
//
//  AppShareViewController.m
//  AppShare
//
//  Created by Lukas Boura on 03/04/2020.
//

#import "AppShareViewController.h"
#import "ShareServiceCell.h"
#import "NSBundle+Resources.h"

@interface AppShareViewController () <UITableViewDataSource, UITableViewDelegate>

@property UITableView *tableView;
@property NSMutableArray<ShareServiceCell *> *cells;

@property AppShareRequest *request;
@property AppShareConfiguration *configuration;
@property ShareFinishedCallback callback;

@end

@implementation AppShareViewController

- (instancetype)initWithRequest:(AppShareRequest *)request configuration:(AppShareConfiguration *)configuration callback:(ShareFinishedCallback)callback {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        self.callback = callback;
        self.request = request;
        self.configuration = configuration;

        self.tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
        self.cells = [[NSMutableArray alloc] init];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupNavigationItem];
    [self setupViews];
    [self setupLayout];
}

- (void)setupNavigationItem {
    self.navigationItem.title = self.configuration.titleText;
    if (@available(iOS 11.0, *)) {
        self.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeAlways;
    }
    UIBarButtonItem *closeButton = [[UIBarButtonItem alloc] initWithImage:[UIImage bundleImageNamed:@"close"] style:UIBarButtonItemStylePlain target:self action:@selector(closeButtonAction)];
    closeButton.tintColor = [UIColor systemGrayColor];
    
    self.navigationItem.leftBarButtonItem = closeButton;
}

- (void)setupViews {
    if (@available(iOS 13.0, *)) {
        self.view.backgroundColor = [UIColor systemBackgroundColor];
    } else {
        self.view.backgroundColor = [UIColor whiteColor];
    }
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.backgroundColor = [UIColor clearColor];
    self.tableView.estimatedRowHeight = 58;
    self.tableView.estimatedSectionHeaderHeight = 32;
    
    for (ShareService *service in [ShareService allServices]) {
        if ([service isAvailable]) {
            ShareServiceCell *cell = [[ShareServiceCell alloc] initWithService:service];
            [self.cells addObject:cell];
        }
    }
}

- (void)setupLayout {
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.view addSubview:self.tableView];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:0],
        [self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:0],
        [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:0],
        [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant:0]
    ]];
}

- (void)closeButtonAction {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.cells.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    return self.cells[indexPath.row];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.cells[indexPath.row].service shareRequest:self.request from:self];
    // inform main gui which share was selected
    if (self.callback != nil) {
        self.callback(self.cells[indexPath.row].service);
    }


}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return UITableViewAutomaticDimension;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewAutomaticDimension;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UILabel *label = [[UILabel alloc] init];
    label.text = self.configuration.descriptionText;
    label.font = [UIFont systemFontOfSize:17];
    label.numberOfLines = 0;
    label.translatesAutoresizingMaskIntoConstraints = NO;
    
    UIView *headerView = [[UIView alloc] init];
    [headerView addSubview:label];
    
    [NSLayoutConstraint activateConstraints:@[
        [label.topAnchor constraintEqualToAnchor:headerView.topAnchor constant:12],
        [label.bottomAnchor constraintEqualToAnchor:headerView.bottomAnchor constant:-12],
        [label.leadingAnchor constraintEqualToAnchor:headerView.leadingAnchor constant:16],
        [label.trailingAnchor constraintEqualToAnchor:headerView.trailingAnchor constant:-16]
    ]];
    
    return headerView;
}

@end

File: Sources/AppShare/AppShareViewController.swift
//
//  AppShareViewController.swift
//  AppShare
//
//  Created by Lukas Boura on 31/03/2020.
//

import UIKit

class AppShareViewController: UIViewController {
    
    // MARK: Views
    
    private let tableView = UITableView(frame: .zero, style: .grouped)
    private var cells: [ShareServiceCell] = [] {
        didSet {
            tableView.reloadData()
        }
    }
    
    // MARK: Dependencies
    
    private let request: AppShareRequest
    
    // MARK: Initialization
    
    init(request: AppShareRequest) {
        self.request = request
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: Life cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupNavigationItem()
        setupViews()
        setupLayout()
    }
    
}

// MARK: - Private API

private extension AppShareViewController {
    
    // MARK: Setup
    
    func setupNavigationItem() {
        navigationItem.title = "Share app"
        if #available(iOS 11.0, *) {
            navigationItem.largeTitleDisplayMode = .always
        }
        let closeButton = UIBarButtonItem(
            image: .bundleImage(named: "close"),
            style: .plain,
            target: self,
            action: #selector(closeButtonAction)
        )
        if #available(iOS 13.0, *) {
            closeButton.tintColor = .systemGray
        } else {
            closeButton.tintColor = .gray
        }
        navigationItem.leftBarButtonItem = closeButton
    }
    
    func setupViews() {
        if #available(iOS 13.0, *) {
            view.backgroundColor = .systemBackground
        } else {
            view.backgroundColor = .white
        }
        
        tableView.delegate = self
        tableView.dataSource = self
        tableView.separatorStyle = .none
        tableView.backgroundColor = .clear
        
        cells = ShareService.allCases
            .filter(\.isAvailable)
            .map(ShareServiceCell.init)
    }
    
    func setupLayout() {
        tableView.translatesAutoresizingMaskIntoConstraints = false

        view.addSubview(tableView)
        
        NSLayoutConstraint.activate([
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 0),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: 0),
            tableView.topAnchor.constraint(equalTo: view.topAnchor, constant: 0),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: 0)
        ])
    }
    
    // MARK: Action
    
    @objc func closeButtonAction() {
        dismiss(animated: true)
    }
    
}

extension AppShareViewController: UITableViewDelegate, UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return cells.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return cells[indexPath.row]
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        cells[indexPath.row].service.share(request, from: self)
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let label = UILabel()
        label.text = "Recommend the app to your friends"
        label.font = .systemFont(ofSize: 17)
        label.numberOfLines = 0
        label.translatesAutoresizingMaskIntoConstraints = false
        
        let headerView = UIView()
        headerView.addSubview(label)
        
        NSLayoutConstraint.activate([
            label.topAnchor.constraint(equalTo: headerView.topAnchor, constant: 12),
            label.bottomAnchor.constraint(equalTo: headerView.bottomAnchor, constant: -12),
            label.leadingAnchor.constraint(equalTo: headerView.leadingAnchor, constant: 16),
            label.trailingAnchor.constraint(equalTo: headerView.trailingAnchor, constant: -16)
        ])
        
        return headerView
    }
    
}
